runhouse/rns/secrets.py

```diff
@@ -146,100 +156,122 @@ def get(
+            logger.info(f"Saving secrets for {provider_name} to local config")
             p.save_secrets(secrets)
 
         return secrets
 
     @classmethod
-    def delete(cls, providers: List[str]):
-        """Delete secrets from Vault for the specified providers"""
+    def delete_from_vault(cls, providers: Optional[List[str]] = None):
+        """Delete secrets from Vault for specified providers (builtin or custom).
+        If none are provided, will delete secrets for all providers which have been enabled in the local environment."""
+        providers = providers or cls.builtin_providers(as_str=True)
         for provider in providers:
-            provider_cls_name = cls.provider_cls_name(provider)
-            p = cls.get_class_from_name(provider_cls_name)
-            if p is None:
-                continue
-
-            p.delete_secrets_from_vault()
-            logger.info(f"Successfully deleted {provider} secrets from Vault")
+            url = f"{rns_client.api_server_url}/{cls.USER_ENDPOINT}/{provider}"
+            resp = requests.delete(url, headers=rns_client.request_headers)
+            if resp.status_code != 200:
+                logger.error(
+                    f"Failed to delete secrets from Vault: {json.loads(resp.content)}"
+                )
 
     @classmethod
-    def delete_secrets_from_vault(cls):
-        resp = requests.delete(
-            f"{rns_client.api_server_url}/{cls.USER_ENDPOINT}/{cls.PROVIDER_NAME}",
-            headers=rns_client.request_headers,
-        )
-        if resp.status_code != 200:
-            raise Exception(f"Failed to delete {cls.PROVIDER_NAME} secrets from Vault")
+    def to(cls, hardware: Union[str, "Cluster"], providers: Optional[List] = None):
+        """Copy secrets to the desired hardware for a list of builtin providers. If no providers are specified
+        will load all builtin providers that are already enabled."""
+        if isinstance(hardware, str):
+            from runhouse import cluster
+
+            hardware = cluster(name=hardware)
+
+        hardware_name = hardware.name
+        if not hardware.is_up():
+            raise RuntimeError(
+                f"Hardware {hardware_name} is not up. Run `hardware_obj.up()` to re-up the cluster."
+            )
+
+        provider_secrets: list = cls.load_provider_secrets(providers=providers)
+        if not provider_secrets:
+            # If no secrets saved in local config files, check if they exist in Vault
+            vault_secrets: dict = cls.download_into_env(save_locally=False)
+            # TODO [JL] change this API so we don't have to convert the list to a dict?
+            provider_secrets: list = [
+                {"provider": k, **v} for k, v in vault_secrets.items()
+            ]
+
+        # Send provider secrets over RPC to the cluster, then save each provider's secrets into their respective files
+        hardware.add_secrets(provider_secrets)
```

Comment:
Oh I see, this is where we use the RPC. So inside the cluster class, let's just make send_secrets call `Secrets.to(self` so it goes through the same code path. It's also nice that we load from vault if the secrets aren't saved locally, which we don't have in that method right now.